lib/subgraphManagers/IntrospectAddNodeAndCompose.ts

```diff
@@ -0,0 +1,321 @@
+    if (this.config.pollIntervalInMs) {
+      this.beginPolling();
+    }
+
+    if (!initialSupergraphSdl) {
+      throw new NoInitialSdlError();
+    }
+
+    this.completeSupergraphSdl = initialSupergraphSdl;
+
+    return {
+      cleanup: async (): Promise<void> => {
+        if (this.state.phase === 'polling') {
+          await this.state.pollingPromise;
+        }
+        this.state = { phase: 'stopped' };
+        if (this.timerRef) {
+          clearTimeout(this.timerRef);
+          this.timerRef = null;
+        }
+      },
+      supergraphSdl: initialSupergraphSdl,
+    };
+  }
+
+  private async updateSupergraphSdl(): Promise<string | null> {
+    if (!this.originalSubgraphs.length) {
+      throw new NoSubgraphsError();
+    }
+    const result = await loadServicesFromRemoteEndpoint({
+      getServiceIntrospectionHeaders: async service => {
+        return typeof this.config.introspectionHeaders === 'function'
+          ? // eslint-disable-next-line no-return-await
+            await this.config.introspectionHeaders(service)
+          : this.config.introspectionHeaders;
+      },
+      serviceList: this.originalSubgraphs,
+      serviceSdlCache: this.serviceSdlCache,
+    });
+
+    if (!result.isNewSchema) {
+      return this.completeSupergraphSdl;
+    }
+
+    if (!result.serviceDefinitions) {
+      return this.completeSupergraphSdl;
+    }
+
+    const { nodeTypeDefs } = result.serviceDefinitions.reduce<NodeTypes>(
+      (acc, service: ServiceDefinition) => {
+        const newServiceDefinition = visit(service.typeDefs, {
+          ObjectTypeDefinition(node) {
+            const name = node.name.value;
+            if (
+              acc.seenNodeTypes.has(name) ||
+              !IntrospectAddNodeAndCompose.objectImplementsNodeInterface(node)
+            ) {
+              return;
+            }
+            acc.nodeTypeDefs.push(
+              IntrospectAddNodeAndCompose.generateNodeType(name),
+            );
+            acc.seenNodeTypes.add(name);
+          },
+        });
+        acc.serviceDefinitions.push(
+          newServiceDefinition
+            ? { ...service, typeDefs: newServiceDefinition }
+            : service,
+        );
+        return acc;
+      },
+      { nodeTypeDefs: [], seenNodeTypes: new Set(), serviceDefinitions: [] },
+    );
+
+    const {
+      directives = [],
+      nodeQueryDirectives,
+      typeIDDataSource,
+    } = this.nodeGatewayConfig;
+    const nodeSchema = createNodeService(
+      typeIDDataSource,
+      nodeQueryDirectives,
+      directives,
+      nodeTypeDefs,
+    );
+
+    if (!this.getDataSource) {
+      throw new NoGetDataSourceMethodDefinedError();
+    }
+    this.nodeServiceLocalGraphQLDataSource = new LocalGraphQLDataSource(
+      nodeSchema,
+    );
+
+    const nodeServiceDataSource = this.getDataSource({
+      name: NODE_SERVICE_NAME,
+      url: NODE_SERVICE_URL,
+    }) as LocalGraphQLDataSource;
```

Comment:
create another function that calls `getDataSource` and returns it already in the shape `composeServices()` need, see the constants for `name` and `url` are the same, you just need to add `typeDefs: bla.sdl()`, this way `createSupergraph...` just calls `composeServices` directly, all it would do is check errors.